File: samples/oma_mail.daimon.pl
#!/bin/env perl -w
use strict;

use DBI;

################ modify these ##################################################
my %MTA	= ( 'virtual'	=> '/etc/postfix/db/virtual',
	'regexp'	=> '/etc/postfix/db/virtual.regex',
	'domains'	=> '/etc/postfix/db/domains' );
my $passwd_cache	= '/var/lib/pam_mysql.cache';
my $postprocess		= '/usr/sbin/postmap ';

my %DB = (  'TYPE'	=> 'mysql',
	'HOST'	=> 'localhost',
	'USER'	=> 'your-MySQL-User',
	'PASS'	=> '##MysqlSecret-SELECT-only##',
	'DB'	=> 'yourMySQL-DB' );

################ no need for modifying anything below ##########################
my $dbh;
my $sth;

$dbh = DBI->connect('DBI:'.$DB{'TYPE'}.':'.$DB{'DB'}.':'.$DB{'HOST'}, $DB{'USER'}, $DB{'PASS'})
	|| die 'Cannot connect to database.';

if(! -e $MTA{'virtual'} || rand(6)<1 || amountOfNewEntries('virtual') > 0) {
	writeFile('virtual', 'virtual', 'address', 'dest', 'address DESC');
}
if(! -e $MTA{'regexp'} || rand(6)<1 || amountOfNewEntries('virtual_regexp') > 0) {
	writeFile('regexp', 'virtual_regexp', 'reg_exp', 'dest', 'LENGTH(reg_exp) DESC');
}
if(! -e $MTA{'domains'} || rand(48)<1) {
	writeDomains();
}
if(defined $passwd_cache && (!(-e $passwd_cache) || rand(24)<1)) {
	writePasswdCache();
}

if(rand(25) < 1) {
	$dbh->do(q{OPTIMIZE TABLE virtual, virtual_regexp, user, domains});
}

$dbh->disconnect;

################################################################################
sub writePasswdCache {
################################################################################
	my @row;

	$sth = $dbh->prepare(q{
		SELECT mbox, pass_crypt
		FROM user
		});
	$sth->execute();

	open(OUT, '>', $passwd_cache);
	while(@row = $sth->fetchrow_array()) {
		print OUT $row[0].':'.$row[1]."\n";
	}
	close OUT;
}

################################################################################
sub writeDomains {
################################################################################
	my @row;

	$sth = $dbh->prepare(q{
		SELECT domain
		FROM domains
		ORDER BY (SELECT count(*)
			  FROM virtual
			  WHERE address LIKE CONCAT('%@', domain))
			 DESC
		});
	$sth->execute();

	open(OUT, '>', $MTA{'domains'});
	while(@row = $sth->fetchrow_array()) {
		print OUT $row[0]."\t\t".$row[0]."\n";
	}
	close OUT;
	system($postprocess.$MTA{'domains'});
}

################################################################################
sub writeFile {
################################################################################
	my @row;

	$sth = $dbh->prepare(qq{
		SELECT $_[2], $_[3]
		FROM $_[1]
		WHERE active=1 ORDER BY $_[4]
		});
	$sth->execute();

	open(OUT, '>', $MTA{$_[0]});
	while(@row = $sth->fetchrow_array()) {
		print OUT $row[0]."\t\t".$row[1]."\n";
	}
	close OUT;

	$dbh->do(qq{UPDATE LOW_PRIORITY IGNORE $_[1] SET neu=0 WHERE neu=1});
	system($postprocess.$MTA{$_[0]});
}

################################################################################
sub amountOfNewEntries {
################################################################################
	my $row;

	$sth = $dbh->prepare(qq{
		SELECT COUNT(*) AS neue
		FROM $_[0]
		WHERE neu=1
		});
	$sth->execute();

	$row = $sth->fetchrow_hashref();
	$sth->finish();
	return $row->{'neue'};
}

File: samples/oma_mail.daimon.php
#!/bin/env php
<?php
/* *********************** change these according to your environment ******* */
$MTA['virtual']	= '/etc/postfix/db/virtual';
$MTA['regexp']	= '/etc/postfix/db/virtual.regex';
$MTA['domains']	= '/etc/postfix/db/domains';
$POSTPROCESS	= '/usr/sbin/postmap %s';
// Set this to null if you don't use pam_pwdfile for caching.
$PASSWD_CACHE	= '/var/lib/pam_mysql.cache';

$DB = array(
	'DSN'		=> 'mysqli://User:Passwd@localhost/DB',
	'PREFIX'	=> '',
);

/* *********************** functions **************************************** */
/* *********************** don't edit anything below +++++++++++++++++******* */
function make_hashfile_of_query($file, $query, $delimiter = "\t\t", $postprocess = true) {
	global $POSTPROCESS, $db;

	if($fp = fopen($file, 'w')) {
		$result = $db->Execute($query);
		while(!$result->EOF) {
			fputs($fp, $result->fields[0].$delimiter.$result->fields[1]."\n");
			$result->MoveNext();
		}
		fclose($fp);

		if($postprocess) {
			exec(sprintf($POSTPROCESS, $file));
		}
	}
}

/* *********************** logic ******************************************** */
include('adodb/adodb.inc.php');
$db	= ADONewConnection($DB['DSN']);
$db->SetFetchMode(ADODB_FETCH_NUM);

// virtual
$amount_new	= $db->GetOne('SELECT COUNT(*) FROM '.$DB['PREFIX'].'virtual WHERE neu=1');
if(!is_file($MTA['virtual']) || $amount_new > 0 || time()%6<1) {
	make_hashfile_of_query($MTA['virtual'], 'SELECT address,dest FROM '.$DB['PREFIX'].'virtual WHERE active=1 ORDER BY address DESC');
	$db->Execute('UPDATE '.$DB['PREFIX'].'virtual SET neu=0 WHERE neu=1');
}

// virtual.regexp
$amount_new	= $db->GetOne('SELECT COUNT(*) FROM '.$DB['PREFIX'].'virtual_regexp WHERE neu=1');
if(!is_file($MTA['regexp']) || $amount_new > 0 || time()%6<1) {
	make_hashfile_of_query($MTA['regexp'], 'SELECT reg_exp,dest FROM '.$DB['PREFIX'].'virtual_regexp WHERE active=1 ORDER BY LENGTH(reg_exp) DESC', "\t\t", false);
	$db->Execute('UPDATE '.$DB['PREFIX'].'virtual_regexp SET neu=0 WHERE neu=1');
}

// domains
$amount_new	= $db->GetOne('SELECT COUNT(*) FROM '.$DB['PREFIX'].'domains WHERE neu=1');
if(!is_file($MTA['domains']) || $amount_new > 0 || time()%96<1) {
	make_hashfile_of_query($MTA['domains'], 'SELECT domain,domain FROM '.$DB['PREFIX'].'domains');
	$db->Execute('UPDATE '.$DB['PREFIX'].'domains SET neu=0 WHERE neu=1');
}

// for pam_pwdfile
if(!is_null($PASSWD_CACHE) && (!is_file($PASSWD_CACHE) || time()%24<1)) {
	make_hashfile_of_query($PASSWD_CACHE, 'SELECT mbox, password FROM '.$DB['PREFIX'].'user', ':', false);
}

// optimize DB
if(time()%50 < 2 && ($DB['TYPE'] == 'mysql' || $DB['TYPE'] == 'mysqli')) {
	$db->Execute('OPTIMIZE TABLE '.$DB['PREFIX'].'virtual, '.$DB['PREFIX'].'virtual_regexp, '.$DB['PREFIX'].'user, '.$DB['PREFIX'].'domains');
}
?>